src/Workspaces/Core/Portable/Workspace/Solution/TextLoader.cs

```diff
@@ -31,9 +38,19 @@ public abstract class TextLoader
         /// <exception cref="InvalidDataException"/>
         /// <exception cref="OperationCanceledException"/>
         public virtual Task<TextAndVersion> LoadTextAndVersionAsync(CancellationToken cancellationToken)
+        {
 #pragma warning disable CS0618 // Type or member is obsolete
-            => LoadTextAndVersionAsync(workspace: null, documentId: null, cancellationToken);
-#pragma warning restore CS0618
+            if (ImmutableInterlocked.GetOrAdd(
+                ref s_isObsoleteLoadTextAndVersionAsyncOverriden,
```

Comment:
Not sure if we even need the cache at this point.

Reply:
Rather than speculating, I wrote a quick benchmark:

```
|    Method |                Job |            Runtime | Checking |      Mean |    Error |    StdDev |
|---------- |------------------- |------------------- |--------- |----------:|---------:|----------:|
| Precached |           .NET 6.0 |           .NET 6.0 |     Base |  24.04 ns | 0.495 ns |  0.741 ns |
|  Uncached |           .NET 6.0 |           .NET 6.0 |     Base | 286.38 ns | 5.652 ns | 10.616 ns |
|   NoCache |           .NET 6.0 |           .NET 6.0 |     Base | 134.84 ns | 2.663 ns |  2.735 ns |
| Precached | .NET Framework 4.8 | .NET Framework 4.8 |     Base |  42.07 ns | 0.844 ns |  1.127 ns |
|  Uncached | .NET Framework 4.8 | .NET Framework 4.8 |     Base | 485.56 ns | 9.590 ns | 13.127 ns |
|   NoCache | .NET Framework 4.8 | .NET Framework 4.8 |     Base | 213.21 ns | 3.195 ns |  2.833 ns |
| Precached |           .NET 6.0 |           .NET 6.0 |  Derived |  21.72 ns | 0.197 ns |  0.175 ns |
|  Uncached |           .NET 6.0 |           .NET 6.0 |  Derived | 282.95 ns | 5.548 ns |  5.189 ns |
|   NoCache |           .NET 6.0 |           .NET 6.0 |  Derived | 137.95 ns | 2.708 ns |  2.533 ns |
| Precached | .NET Framework 4.8 | .NET Framework 4.8 |  Derived |  40.03 ns | 0.202 ns |  0.179 ns |
|  Uncached | .NET Framework 4.8 | .NET Framework 4.8 |  Derived | 478.76 ns | 5.739 ns |  5.368 ns |
|   NoCache | .NET Framework 4.8 | .NET Framework 4.8 |  Derived | 220.71 ns | 0.819 ns |  0.684 ns |
```

The first `GetOrAdd` takes approximately double the time the uncached version takes, but subsequent accesses on the same type are at least 5 times faster (though there is no parallelism in my benchmark, so these are best-case scenarios for `GetOrAdd`). Assuming that we'll see a very small number of implementations of this class, I think the cache makes sense.

Benchmark code:
<details>
```cs
using System.Collections.Immutable;
using System.Reflection;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

BenchmarkRunner.Run<Benchmark>();

[SimpleJob(BenchmarkDotNet.Jobs.RuntimeMoniker.Net60)]
[SimpleJob(BenchmarkDotNet.Jobs.RuntimeMoniker.Net48)]
public class Benchmark
{
    private ImmutableDictionary<Type, bool> _precached = ImmutableDictionary<Type, bool>.Empty;
    private ImmutableDictionary<Type, bool> _benchmarkFirstAdd = ImmutableDictionary<Type, bool>.Empty;

    [Params(typeof(Base), typeof(Derived))]
    public Type Checking = null!;

    private Base _base = null!;

    [GlobalSetup]
    public void Setup()
    {
        _precached = _precached.Add(typeof(Base), false);
        _precached = _precached.Add(typeof(Derived), true);
        _base = Checking == typeof(Base) ? new Base() : new Derived();
    }

    [Benchmark]
    public bool Precached()
    {
        _benchmarkFirstAdd = ImmutableDictionary<Type, bool>.Empty;
        return ImmutableInterlocked.GetOrAdd(ref _precached, Checking, _ => _base.Check());
    }

    [Benchmark]
    public bool Uncached()
    {
        // Added to all because we're benchmarking a very fast operation, and BenchmarkDotNet will automatically
        // put multiple iterations in the same run to remove noise. IterationSetup removes the ability to do this,
        // so it's easier to just add the reset assignment to every benchmark and let additional load and assignment
        // be added as a constant to all benchmark runs.
        _benchmarkFirstAdd = ImmutableDictionary<Type, bool>.Empty;
        return ImmutableInterlocked.GetOrAdd(ref _benchmarkFirstAdd, Checking, _ => _base.Check());
    }

    [Benchmark]
    public bool NoCache()
    {
        _benchmarkFirstAdd = ImmutableDictionary<Type, bool>.Empty;
        return _base.Check();
    }
}

internal class Base
{
    public virtual void M() { }

    public bool Check()
    {
        return new Action(M).GetMethodInfo().DeclaringType == typeof(Base);
    }
}

internal class Derived : Base
{
    public override void M() { }
}
```
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFrameworks>net6.0;net48</TargetFrameworks>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <LangVersion>latest</LangVersion>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="BenchmarkDotNet" Version="0.13.2" />
  </ItemGroup>

</Project>
```
</details>